### DISCUSSION TITLE
PCPCombine "subtract" method using a bucket interval?

### DISCUSSION THREAD

**Discussion by MarcelCaron-NOAA** [dtcenter/METplus#1812] [Configuration] [answered]
I’d like to use PCPCombine to make 1-h precipitation accumulations.  Of the forecast files in my input dataset, some forecasts already have 1-h precipitation fields, but the remaining files have either 3-h (A3) or 2-h (A2) fields, in which case I’d like to “subtract” the, for example, A3 field with the A2 field in the previous forecast.  

I can do this by running PCPCombine on the command line, where it is possible to select A3 as the level for the first file and A2 for the second using the following command (example for NCEP's WCOSS2 computer):

`/apps/ops/para/libs/intel/19.1.3.304/met/10.1.1/bin/pcp_combine -v 5 -subtract /lfs/h1/ops/prod/com/nam/v4.2/nam.20220913/nam.t00z.alaskanest.hiresf06.tm00.grib2 'name="APCP"; level="A03";' /lfs/h1/ops/prod/com/nam/v4.2/nam.20220913/nam.t00z.alaskanest.hiresf05.tm00.grib2 'name="APCP"; level="A02";' /lfs/h2/emc/stmp/$USER/tmp/namnest.t00z.f006.a01h.ak.nc`

 ... but I wonder if it is also possible to have METplus do the same using a configuration file?  

Currently, I have the following settings in my PCPCombine configuration file...: 

```
FCST_PCP_COMBINE_METHOD = SUBTRACT
FCST_PCP_COMBINE_INPUT_ACCUMS = {lead}
FCST_PCP_COMBINE_OUTPUT_ACCUM = 01
FCST_PCP_COMBINE_BUCKET_INTERVAL = 3 
```

... because the accumulation field in each forecast resets every 3 lead hours for this particular dataset.  However, METplus seems to ignore the bucket interval, and instead uses the full lead time as the accumulation (e.g., for F06, METplus uses APCP/A6 - APCP/A5, when I'd like it to use APCP/A3 - APCP/A2).

Is it possible for METplus to reset the accumulation based on the bucket interval and lead time for a "subtract" PCPCombine run?

----------

Attached are a log file and a configuration file, as well as a list of the environment variables used to set some of the METplus configurations for a sample run.

[LOG_PCPCombine_fcstnamnest_precip_mrms_job33_for2022091310_runon20220916175343.log](https://github.com/dtcenter/METplus/files/9586746/LOG_PCPCombine_fcstnamnest_precip_mrms_job33_for2022091310_runon20220916175343.log)
[CONF_PCPCombine_fcstnamnest_precip_mrms_job33_for2022091310_runon20220916175343.txt](https://github.com/dtcenter/METplus/files/9586747/CONF_PCPCombine_fcstnamnest_precip_mrms_job33_for2022091310_runon20220916175343.txt)
[ENV_VARS_job33.txt](https://github.com/dtcenter/METplus/files/9586743/ENV_VARS_job33.txt)

**Comment by georgemccabe** [comment]
Hi @MarcelCaron-NOAA, currently the bucket interval logic is only applied to PCPCombine using ADD or DERIVE methods. It is not supported for the SUBTRACT method.

**Reply by MarcelCaron-NOAA** [reply]
Hi @georgemccabe, thanks for the information.  As a follow-up question, is it possible to define via a configuration file the two fields that need to be used in the subtraction?  (See excerpt below)

> I can do this by running PCPCombine on the command line, where it is possible to select A3 as the level for the first file and A2 for the second using the following command (example for NCEP's WCOSS2 computer):

>/apps/ops/para/libs/intel/19.1.3.304/met/10.1.1/bin/pcp_combine -v 5 -subtract /lfs/h1/ops/prod/com/nam/v4.2/nam.20220913/nam.t00z.alaskanest.hiresf06.tm00.grib2 'name="APCP"; level="A03";' /lfs/h1/ops/prod/com/nam/v4.2/nam.20220913/nam.t00z.alaskanest.hiresf05.tm00.grib2 'name="APCP"; level="A02";' /lfs/h2/emc/stmp/$USER/tmp/namnest.t00z.f006.a01h.ak.nc

>... but I wonder if it is also possible to have METplus do the same using a configuration file?

**Comment by georgemccabe** [comment, marked as answer]
Hi @MarcelCaron-NOAA,

The pcp_combine tool does not utilize a configuration file at all. However, if you know that the A03 and A02 levels will be in your files, you can configure the PCPCombine wrapper using the USER_DEFINED method, which allows you to explicitly specify the command that you want to generate for each run time. The [Shift Keyword](https://metplus.readthedocs.io/en/develop/Users_Guide/systemconfiguration.html?highlight=truncate#shift-keyword) can be used to adjust the time for the 2nd input file. You could set something like this:

```
FCST_PCP_COMBINE_METHOD = USER_DEFINED

FCST_PCP_COMBINE_COMMAND = -subtract /lfs/h1/ops/prod/com/nam/v4.2/nam.{init?fmt=%Y%m%d}/nam.t{init?fmt=%H}z.alaskanest.hiresf{lead?fmt=%2H}.tm00.grib2 'name="APCP"; level="A03";' /lfs/h1/ops/prod/com/nam/v4.2/nam.{init?fmt=%Y%m%d}/nam.t{init?fmt=%H}z.alaskanest.hiresf{lead?fmt=%2H?shift=-1H}.tm00.grib2 'name="APCP"; level="A02";'
```

The value of FCST_PCP_COMBINE_COMMAND is substituted and added to the call to pcp_combine. You do not need to include the pcp_combine executable, the verbosity level (set with -v), and the output filename. The output filename is automatically added to the end of the command using FCST_PCP_COMBINE_OUTPUT_DIR/TEMPLATE.

For valid time 2022-09-13 10Z and forecast lead 4 hour, the command that is generated from this example would look something like this:

> pcp_combine -v 2 -subtract /lfs/h1/ops/prod/com/nam/v4.2/nam.20220913/nam.t06z.alaskanest.hiresf04.tm00.grib2 'name="APCP"; level="A03";' /lfs/h1/ops/prod/com/nam/v4.2/nam.20220913/nam.t06z.alaskanest.hiresf03.tm00.grib2 'name="APCP"; level="A02";' /path/to/output/file.nc

Here is an example use case that uses [PCPCombine USER_DEFINED method](https://metplus.readthedocs.io/en/develop/generated/met_tool_wrapper/PCPCombine/PCPCombine_user_defined.html?highlight=USER_DEFINED).

Let me know if that will accomplish what you need.

**Reply by MarcelCaron-NOAA** [reply]
@georgemccabe 

Yes the USER-DEFINED method with the "shift" keyword works excellently.  Thanks for the help!
